src/Microsoft.VisualStudio.Threading/AsyncAutoResetEvent.cs

```diff
@@ -99,7 +99,15 @@ public Task WaitAsync(CancellationToken cancellationToken)
                 else
                 {
                     var waiter = new WaiterCompletionSource(this, cancellationToken, this.allowInliningAwaiters);
-                    this.signalAwaiters.Enqueue(waiter);
+                    if (cancellationToken.IsCancellationRequested)
```

Comment:
There is still a race: what if on line 102 the cancellation is not yet requested, and on line 108 it has already been requested, the callback has had enough time to fire and it couldn't find the ```waiter``` in ```this.signalAwaiters```?

Reply:
A good question, but not a race condition, because this code is in a `lock (this.signalAwaiters)` block, and so is the cancellation registration callback, so they can't run concurrently.

Reply:
(I verified in a debugger by carefully controlling thread execution order and I could not cause the interleaving you proposed because of the locks taken).